### PR TITLE
Add security considerations text around "claims"

### PR DESCRIPTION
**Pull request by evan2645**
This commit updates text in the SPIFFE ID specification to address the
common question of supporting "claims" in SVIDs. The text attempts to
explain some of the gotchas associated with "claim" use (which we refer
to here as "assertions"), and provides guidelines by which readers may
assess the relative safety of any given assertion.

This text has been authored by SIG-Spec and was ported from the Google
Doc link below. It supersedes #185.

I hijacked the conclusion section since it was factually questionable and
was adding very little value in my opinion.

https://docs.google.com/document/d/1gtcQ73Cc24XA6uBgeMCS9IOcUEvYpbj8iTiUiuMFtvE

Signed-off-by: Evan Gilman <egilman@vmware.com>

### CODE REVIEW

File: standards/SPIFFE-ID.md
Comment (spikecurtis) [inline standards/SPIFFE-ID.md]: > Assertions that are directly formalized by the SPIFFE specification are considered safe as they have been closely scrutinized.

I think this promises too much.  What does "safe" mean here?  We just got finished saying veracity of the assertion is a concern, so are we promising that no one will ever lie in any SPIFFE-formalized assertion?

File: standards/SPIFFE-ID.md
Comment (spikecurtis) [inline standards/SPIFFE-ID.md]: Can we drop "extreme" from the last sentence?

I don't think it is extreme to consider a case where you might federate with a trust domain and trust them to make certain kinds of assertions but not others.

File: standards/SPIFFE-ID.md
Comment (spikecurtis) [inline standards/SPIFFE-ID.md]: This example is more charitable to trust domain B than I think is necessary in that it doesn't mention the idea that someone in trust domain B might be lying.

File: standards/SPIFFE-ID.md
Comment (spikecurtis) [inline standards/SPIFFE-ID.md]: I don't think "that is not explicitly defined by the SPIFFE specification set" is the right qualification here.

The question administrators need to ask themselves is just a blanket, "what assertions do I trust this domain to make?"  They should NOT just assume that SPIFFE-defined assertions are true!  The difference between SPIFFE-defined and not is in how an operator should react.  If the operator does not trust SPIFFE-defined assertions, then maybe they should not federate at all.  If they trust SPIFFE-defined assertions, but not some other assertions, then it might be fine to federate and just filter out those other assertions.

**Review by spikecurtis — CHANGES_REQUESTED** [review]

File: standards/SPIFFE-ID.md
Comment (amoore877) [reply standards/SPIFFE-ID.md:152]: In what cases would we not trust another domain's issued SPIFFE IDs (for instance) but still say that we trust the domain at all? I guess maybe what the intention here is is to say there is a difference between "core" attributes which that come with "I trust this domain" vs extra attributes that come with considerations and allowlists/blocklists.

Comment (amoore877) [reply standards/SPIFFE-ID.md]: I think certain trust assumptions come with trusting a domain, the core one being ID. If you do not trust the IDs coming out of a domain, then I don't think you can trust anything coming from the domain.

So there is "safety" in such core attributes as long as due diligence (is trusting them necessary, are security practices good...) has been performed in trusting the domain itself.
Other attributes that a domain decides to add are potentially suspect and need the further examination being documented.

File: standards/SPIFFE-ID.md
Comment (amoore877) [reply standards/SPIFFE-ID.md]: IMO "someone in trust domain B is maliciously using the authority" is a subproblem of "trust domain B is compromised". Someone or something has made the signing authority sign something it shouldn't.

I'm not sure how much value is added to the example by exhaustively stating insider vs outsider vs accident vs. malicious etc dimensions to the problem.

However I also don't know that we totally need the full example either. We have a business decision where we trust A to tell us about related customer, and we don't have that same decision with B. Allowing B's processes the same access is overpermissioning, regardless of if such a caller is ever actually present.

File: standards/SPIFFE-ID.md
Comment (spikecurtis) [reply standards/SPIFFE-ID.md]: > I think certain trust assumptions come with trusting a domain, the core one being ID. If you do not trust the IDs coming out of a domain, then I don't think you can trust anything coming from the domain.

I agree with this statement, but I think it is worth stating explicitly, rather than the current language.

Comment (evan2645) [reply standards/SPIFFE-ID.md]: I feel like the SPIFFE ID is a bit of a special case. In many ways, its value is completely arbitrary and does not matter when considering the loosely coupled federation use cases.

For example, when federating with a vendor that you send data to for indexing, the business has already made a decision to trust this vendor (the entirety of this vendor - the legal entity) with its data. Nobody in this business is saying they trust this data to a specific workload running within the vendor, instead the domain of trust is the vendor as a whole. But, we still probably want to validate the vendor's SPIFFE ID (beyond trust domain) right?

I think the reality is that this validation is a quick and easy win that will probably make a measurable improvement .. however if the SPIFFE ID is lied about (by the vendor, to be the "correct" one), it doesn't really matter because the net effect is the same, and the trust expectations of the business have not been violated.

So I guess the argument I'm trying to make here is that, maybe "trusting the SPIFFE ID" is a moot point because even if it is not trustworthy, lying about it or otherwise does not actually violate any of the models I can think of? It is effectively equivalent to a trust domain compromise at that point.

I have more follow-on thoughts that will bring me back to the original statement questioned in this comment, but I think I'll stop short of there because I'd like to see what people think of the above first

Comment (amoore877) [reply standards/SPIFFE-ID.md]: > So I guess the argument I'm trying to make here is that, maybe "trusting the SPIFFE ID" is a moot point because even if it is not trustworthy, lying about it or otherwise does not actually violate any of the models I can think of? It is effectively equivalent to a trust domain compromise at that point.

Thinking from the perspective here of "this other trust domain is an entirely different business I have no view into" for me adds clarity to the wording. We have a business agreement with `B`, and a technical implementation agreement with `B` that `spiffe://b/workload/foo` will get access to our internal resource `bar`. How `B` engineers decide internally what qualifies as `spiffe://b/workload/foo` in standard operations, experimentation, mitigation scenarios, total hacks, or migrations is entirely up to them and we have no view of whether the `B` engineers are "lying" or not because we don't know their operations to begin with. Nor do we care about the gritty internal operations because our agreement is simply "we will allow `spiffe://b/workload/foo`, and you will follow as best of practices as you can to ensure integrity of issuance of `spiffe://b/workload/foo`".

Disclosed / public security events of course alter the scenario, but on the business level (eg. "`B`, you need to address these CVEs if we're going to continue doing business with each other").

Comment (spikecurtis) [reply standards/SPIFFE-ID.md]: I mean, either you care which SPIFFE-ID is on the other end of the line, or you don't.  If you don't care, it doesn't make any sense to validate it, and you might still decide to federate with the domain if all you want is a sort of organization-level mutual-auth. (Or, one side might care while the other does not.)

However, I don't think we should assume that is it generally the case that you don't care about the SPIFFE-ID for federation use cases.  If SPIFFE is widely deployed in an organization then there could be many different applications in the same trust domain.  

To make a real world analogy, let's say you're working with an accounting firm, and you give highly confidential revenue and earning statements to your accountant, Alicia.  One day when you walk in to drop them off, Steve from the cafeteria meets you in the lobby and says Susan is out today, but you should leave the documents with him instead.  Steve has an authentic ID badge issued by the company.  Maybe you've even seen him in the cafeteria slicing bagels, so you have no doubt he does actually work for the accounting firm.  Do you leave the highly sensitive documents with him?  Does the fact that Steve has done this translate into "the accounting firm is compromised" in your mind?

I think we may have in our mind very near-term use cases where SPIFFE control planes are managed by a single application team, or possibly an infrastructure team that serves a small group of related applications.  But thinking beyond that, it might be desirable for there to be a single, centrally managed SPIFFE control plane for a large multinational.  Maybe it even has self-service to allow your little prototype application to get a SPIFFE-ID in some subspace of the  domain SPIFFE-ID space.

Comment (evan2645) [reply standards/SPIFFE-ID.md]: We had a discussion on this today in the SIG-Spec call:

- Everyone agrees that we should not be telling our users to blindly trust the veracity of any given SPIFFE ID "assertion"
- We discussed the fact that trustworthiness of the SPIFFE ID "assertion" is transcended by the trustworthiness of the entity with which you are federating
- Everyone agrees that just because you consider the entity you are federating with to be trustworthy, you should still validate the SPIFFE ID
- If you do not find the entity to be trustworthy, and therefor do not find the SPIFFE ID to be trustworthy, then you should probably not be sending your data to this entity to begin with

Which is aligned with previous statements:
>> I think certain trust assumptions come with trusting a domain, the core one being ID. If you do not trust the IDs coming out of a domain, then I don't think you can trust anything coming from the domain.
>
> I agree with this statement, but I think it is worth stating explicitly, rather than the current language.

Though the above can also be adjusted from "trust anything coming from the domain" to "trust sending any data into the domain".

I _think_ we're all actually aligned here, and everyone on the call was amenable to language similar to that in the last bullet point above. That said, after reviewing this paragraph and surrounding text, I think my suggestion is actually to just strike the offending sentence. If one holds even the SPIFFE specified assertions to the same guidelines we've outlined for everything else here, the same conclusions we're discussing now should be reached.

Alternatively, if we want to call this out directly, I think we should have a new paragraph:
```suggestion
This section explores all four areas of concern, and provides guidelines by which operators may assess the relative safety of any given SVID assertion. Generally speaking, operators should err on the side of caution and only include assertions in which there is a very high degree of confidence in the safety of the assertion in question.

It should be noted that while assertions which are directly formalized by the SPIFFE specifications are not typically vulnerable to problems related to interpretation and meaning, they can still be vulnerable to problems related to veracity. However, due to the tightly scoped nature of SPIFFE defined assertions, veracity concerns in this regard indicate much larger concerns around the security posture of the trust domain in question, at which point operators should seriuosly question whether or not data should be exchanged with these systems in the first place.
```

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: I had put "extreme" here since suggesting nefarious behavior of your peers is generally something I try to shy away from, and because the "X is compromised" case feels a little bit different than an outright lie (_who_ is lying to me?).

I am happy to strike it though, I don't think it's strictly necessary.

File: standards/SPIFFE-ID.md
Comment (azdagron) [reply standards/SPIFFE-ID.md]: > sending any data into the domain

small nit: I think trust extends to both sending data into and receiving data from a federated trust domain.

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: > However I also don't know that we totally need the full example either. We have a business decision where we trust A to tell us about related customer, and we don't have that same decision with B. Allowing B's processes the same access is overpermissioning, regardless of if such a caller is ever actually present.

I think the real problem here is _who_ you trust to make _which_ assertion. In this case, A should really only trust A to make this kind of user identity assertion. If B tries to make the same assertion, it should be denied. In this way, sure it is a misconfiguration ... however it is an easy one to make/overlook.

This is a very slippery slope in this context. We are saying that maybe we trust the security posture of B enough to trust its SPIFFE IDs and authority... because we are federating. We are further saying that even though we trust B enough to federate, we don't trust it enough to claim that it has authenticated one of our customers and then access their data. And it is this point I was trying to highlight with the example. The veracity decision is not applied equally to all assertions, it must be made on a case-by-case basis.

I think it's an important distinction to make, but if I'm muddying the waters with this example I'd love some suggestions on how to make this clearer.

File: standards/SPIFFE-ID.md
Comment (spikecurtis) [reply standards/SPIFFE-ID.md]: Yeah, just strike it then.

Comment (azdagron) [reply standards/SPIFFE-ID.md]: :+1 to strike

File: standards/SPIFFE-ID.md
Comment (spikecurtis) [reply standards/SPIFFE-ID.md]: I think the overall example is fine; it's just the last sentence reads to me to imply that trust domain B being compromised is the main item in the threat model.  The improvement I would suggest is, instead of giving an example of why you might not want to trust domain B to make this assertion, we instead start from zero trust and ask the reader to consider why they should trust that assertion.  Is there an operational need to trust it?  If so, do they understand how trust domain B is making that assertion?  Does the customer who's data it is agree?

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: I pushed an update that includes the alternate language, which I think Spike may prefer since it directs readers to call federation relationships into question if they feel that they can't trust SPIFFE IDs in a foreign trust domain

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: Stricken

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: I added a small blurb here that B may also have a malicious internal actor.

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: I am worried that we are trending down the highly skeptical security person who always says no and never believes anyone path. I updated this paragraph with the latest commit to exclude the quote you called out, but now the text basically says "don't trust anything in an svid unless you otherwise know it to be likely true". I think this puts our users in a very tough spot, and gives the impression that they should not trust this technology because everything in it is effectively hearsay. It feels too extreme.

**Comment by evan2645** [comment]
hey folks - pushed a new commit to address the outstanding comments. Some have been wholly addressed, some partially addressed. Overall, I am fairly worried about messaging that SPIFFE operators should take an (IMO) impractically suspicious stance when evaluating basic aspects of SPIFFE like the validation of a SPIFFE ID. I certainly think we need to be clear about where the dangers lie, how to identify them and how to mitigate them .. but I also feel we should not be painting a picture of a bad world where you must constantly be on guard, because while this is true for some folks it's not true for most .. and the folks that it _is_ true for, well, they know who they are

File: standards/SPIFFE-ID.md
Comment (amoore877) [reply standards/SPIFFE-ID.md]: I think this goes back to the other comment thread on just being clear about business agreements vs security.
You should never trust anything you don't have to. But if your business A decides to federate with B, now you have to trust them. Your business relationship defines somewhere what elements of each other's SVIDs you will rely on and for what purposes, these elements being things you are otherwise **blindly trusting unless signed by an unknown/untrusted CA, or notified of compromise**, because that is what your business agreement is.

One doesn't even have to go into federation or otherwise separate domains for this concept. For A's own processes to talk to other A processes, the receiver is also just going off of whatever internal agreements there are for usage.

Similar to @evan2645 's concern, IMO in trying to clarify this situation we have made usage seem less and less reliable. For instance, in the example for "a B process says it has authenticated A's customer": in operational usage or permissioning models I wouldn't even consider this something to bring up, because in whatever authorization model I have as a security employee of A I only allow `spiffe://a.ca/customer/authenticator` to state it has authenticated A's customers. And no matter how hard it might try to be otherwise, `spiffe://**b**.ca/customer/authenticator` is not the same identity.

Comment (spikecurtis) [reply standards/SPIFFE-ID.md]: I'm actually happy just dropping this sentence entirely.  In the same section we have this sentence

> Operators should carefully consider, on a case by case basis, whether or not a given foreign authority should be considered trustworthy in the context of a specific assertion.

Which I think conveys the sensible advice we want to give, without the highly skeptical language: "considered generally untrustworthy unless it can be rigorously proven otherwise."

File: standards/SPIFFE-ID.md
Comment (azdagron) [inline standards/SPIFFE-ID.md]: seriuosly => seriously

File: standards/SPIFFE-ID.md
Comment (azdagron) [reply standards/SPIFFE-ID.md]: > I think this puts our users in a very tough spot, and gives the impression that they should not trust this technology because everything in it is effectively hearsay. It feels too extreme.

I agree with @evan2645's concern. I think we want to point out to people that they should scrutinize assertions for trustworthiness without conveying tinfoil hat levels of paranoia. You have to be able to trust _something_.

Accordingly, I would also be happy to drop this sentence.

Comment (evan2645) [reply standards/SPIFFE-ID.md]: Ok great, I think that works well in resolving my concern. It has been stricken

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: Fixed

**Review by azdagron — APPROVED** [review]
Looks good to me. Thanks for the discussion on this, everybody.

**Comment by evan2645** [comment]
I have some changes staged on disk to address the follow-up on #183 

Those changes add another security considerations subsection and are built on top of this PR, so I am holding them back until this one is merged. Waiting for @spikecurtis's approval, it would also be good to get @amoore877 to have one last look.

**Comment by amoore877** [comment]
lgtm, though I don't have an Approval option

**Review by spikecurtis — APPROVED** [review]
lgtm

**Pull request merged**
